<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="date" content="2018-07-19" />

<title>gren: Adaptive Group-Regularized Logistic Elastic Net Regression</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">gren: Adaptive Group-Regularized Logistic Elastic Net Regression</h1>
<h4 class="author"><em>Magnus MÃ¼nch</em></h4>
<address class="author_afil">
<ol style="list-style-type: decimal">
<li>Department of Epidemiology &amp; Biostatistics, VU University Medical Center, Amsterdam, The Netherlands 2. Mathematical Institute, Leiden University, Leiden, The Netherlands</li>
</ol><br><a class="author_email" href="mailto:#"><a href="mailto:m.munch@vumc.nl">m.munch@vumc.nl</a></a>
</address>
<h4 class="date"><em>2018-07-19</em></h4>


<div id="TOC">
<ul>
<li><a href="#overview"><span class="toc-section-number">1</span> Overview</a><ul>
<li><a href="#elements-of-the-package"><span class="toc-section-number">1.1</span> Elements of the package</a></li>
<li><a href="#getting-started"><span class="toc-section-number">1.2</span> Getting started</a></li>
</ul></li>
<li><a href="#example-1-micrornas-in-colon-cancer"><span class="toc-section-number">2</span> Example 1: microRNAs in colon cancer</a><ul>
<li><a href="#data"><span class="toc-section-number">2.1</span> Data</a></li>
<li><a href="#estimation"><span class="toc-section-number">2.2</span> Estimation</a></li>
<li><a href="#interpreting-results"><span class="toc-section-number">2.3</span> Interpreting results</a></li>
<li><a href="#predictive-performance"><span class="toc-section-number">2.4</span> Predictive performance</a></li>
</ul></li>
<li><a href="#example-2-micrornas-in-cervical-cancer"><span class="toc-section-number">3</span> Example 2: microRNAs in cervical cancer</a><ul>
<li><a href="#data-1"><span class="toc-section-number">3.1</span> Data</a></li>
<li><a href="#two-partitions"><span class="toc-section-number">3.2</span> Two partitions</a></li>
</ul></li>
<li><a href="#references">References</a></li>
</ul>
</div>

<div id="overview" class="section level1">
<h1><span class="header-section-number">1</span> Overview</h1>
<p>This document contains the vignette for the R package <code>gren</code>. The document demonstrates the functionality and usage of the package with two example data sets. Also included are some questions regarding the functionality, which require the user to put some more thought into the data analysis. The vignette can also be used when ignoring the questions. All code to answer the questions is included, but for the the sake of computational time, the code is not run and the output is not shown. The package is based on the publication <span class="citation">MÃ¼nch et al. (<a href="#ref-munch_adaptive_2018">2018</a>)</span>.</p>
<p>The package enables the user to enhance classification performance and variable selection of the logistic elastic net through the inclusion of external information. The information comes in the form of a partitioning of the features into non-overlapping group. Example of such co-data are:</p>
<ol style="list-style-type: decimal">
<li>Response-independent summaries in the primary data (e.g.Â standard deviation).</li>
<li>Feature groupings from prior knowledge (e.g.Â pathways from public data bases).</li>
<li>Feature-specific summaries from an independent study (e.g.Â p-values).</li>
</ol>
<div id="elements-of-the-package" class="section level2">
<h2><span class="header-section-number">1.1</span> Elements of the package</h2>
<p>Key elements of the <code>gren</code> package are:</p>
<ol style="list-style-type: decimal">
<li><code>gren</code>: main function of the package that estimated the model.</li>
<li><code>cv.gren</code>: convience function that gives cross-validated predictions</li>
<li><code>coef.gren</code>, <code>predict.gren</code>: extract coefficients and create predictions from an estimated model.</li>
<li><code>denet</code>, <code>renet</code>: density and random number generator for the elastic net prior distribution.</li>
</ol>
</div>
<div id="getting-started" class="section level2">
<h2><span class="header-section-number">1.2</span> Getting started</h2>
<p>The <code>gren</code> package is available online. Installation from CRAN is as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;gren&quot;</span>)</code></pre></div>
<p>Now we may load the package</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(gren)</code></pre></div>
<p>The <code>gren</code> package imports the following R packages: <code>glmnet</code>, <code>Rcpp</code>, <code>Iso</code>, and <code>pROC</code>.</p>
<p>It might be useful to clear the global environment before starting</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</code></pre></div>
</div>
</div>
<div id="example-1-micrornas-in-colon-cancer" class="section level1">
<h1><span class="header-section-number">2</span> Example 1: microRNAs in colon cancer</h1>
<div id="data" class="section level2">
<h2><span class="header-section-number">2.1</span> Data</h2>
<p>In this part, we will be working with an unpublished (anonymised) dataset containing deep sequenced microRNAs for a cohort of treated colorectal cancer patients. The data is read into the global environment with the following piece of code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## load the data
<span class="kw">data</span>(dataColon)</code></pre></div>
<p>MicroRNAs are small non-coding strands of RNA that are believed to be important in epigenetic functioning. The microRNAs are contained in the <code>mirCol</code> object in the form of a <code>matrix</code>. The rows of the <code>matrix</code> contain the samples, while the columns contain the microRNAs. Run</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(mirCol)</code></pre></div>
<p>to find the dimensions of the object containing the microRNAs. Note that <code>gren</code> requires an <span class="math inline">\(n \times p\)</span> input matrix, with <span class="math inline">\(n\)</span> and <span class="math inline">\(p\)</span> the number of samples and features, respectively.</p>
<p><strong><em>QUESTION</em></strong> <em>How many microRNAs do we have? And what is the number of subjects?</em></p>
<p>The aim of the analysis is to classify treatment response, coded as either non-progressive/remission or progressive. The responses are given in the <code>respCol</code> object in the form of a <code>factor</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(respCol)
<span class="kw">table</span>(respCol)</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>What is the number of non-progressive/remission patients? And the number of progressive patients?</em></p>
<p>In addition there are four clinical covariates that we may include in the analysis. They are contained in the <code>data.frame</code> named <code>unpenCol</code>. the first covariate denotes prior use of adjuvent therapy, a binary variable named <code>adjth</code>. The second is the type of systemic treatment in the form of a ternary variable called <code>thscheme</code>. The third is the <code>numeric</code> vector <code>age</code> and the fourth is the binary variable that codes for primary tumour differentiation called <code>pcrcdiff</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(unpenCol)

## we check distribution of the clinical variables
<span class="kw">apply</span>(unpenCol[, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>)], <span class="dv">2</span>, table)
<span class="kw">hist</span>(unpenCol<span class="op">$</span>age, <span class="dt">main=</span><span class="st">&quot;Distribution of ages&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Age&quot;</span>)</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>How many clinical covariates are there?</em></p>
<p><strong><em>QUESTION</em></strong> <em>Do we want to penalise these clinical covariates? Why/why not?</em></p>
<p>Lastly, there is a <code>factor</code> named <code>mirExpr</code>, which contains the group-memberships of the microRNAs. There are three groups: non differentially expressed microRNAs named <code>nonExpr</code>, medium differentially expressed microRNAS (0.001 <span class="math inline">\(&lt;\)</span> FDR <span class="math inline">\(\leq\)</span> 0.05) named <code>medExpr</code>, and highly differentially expressed microRNAs (FDR <span class="math inline">\(&lt;\)</span> 0.001) called <code>highExpr</code>.These groups are based on a preliminary experiment in a different group of patients, where metastatic colorectal tumour tissue and primary colorectal tissue were compared to normal non-colorectal tissue and primary colorectal tissue, respectively <span class="citation">(Neerincx et al. <a href="#ref-neerincx_mir_2015">2015</a>)</span>. Run the following code to inspect the partitioning:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(mirExpr)
<span class="kw">table</span>(mirExpr)</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>What are the sizes of the different groups of microRNAs?</em></p>
<p>We expect that incorporation of this partitioning enhances therapy response classification, because tumor-specific miRNAs might be more relevant than non-specific ones.</p>
<p><strong><em>QUESTION</em></strong> <em>What group do you expect to receive the largest/smallest penalty?</em></p>
</div>
<div id="estimation" class="section level2">
<h2><span class="header-section-number">2.2</span> Estimation</h2>
<p>After inspection we standardise the microRNA expression levels. Note that the microRNA expression levels have already been pre-processed somewhat, so they are not counts anymore. But since we standardize them anyway, it doesnât really matter.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mirColScaled &lt;-<span class="st"> </span><span class="kw">scale</span>(mirCol)</code></pre></div>
<p>Now that we have looked at the data and standardised the expression levels, we can estimate a group-regularized elastic net model using the <code>gren</code> function. <code>gren</code> takes several arguments, the most important ones are <code>x</code>, <code>y</code>, <code>m</code>, <code>unpenalized</code>, and <code>partitions</code>. These arguments constitute the data used for the estimation. For an explanation of these arguments and others use <code>?gren</code>.</p>
<p>By default, <code>gren</code> estimates a binary logistic regression model. If the outcome data is not binary, but a sum of binary experiments, we may create a <code>numeric</code> vector <code>m</code> that contains the number of Bernoulli trials for each observation. Since our outcome here is binary, we do not specify it here.</p>
<p>In order to estimate the model, we have to specify a penalty mixing parameter <code>alpha</code>. This parameter determines the proportion of <span class="math inline">\(L_1\)</span> and <span class="math inline">\(L_2\)</span>-norm penalty. <code>alpha=0</code> is a ridge model, while <code>alpha=1</code> is the lasso model. We will start with the default value <code>alpha=0.5</code>. If <code>lambda=NULL</code> (default) it is estimated by cross-validation of the regular elastic net model. Furthermore we will include an unpenalized intercept by setting <code>intercept=TRUE</code> (default).</p>
<p><strong><em>QUESTION</em></strong> <em>Can you think of any reason why we donât wish to penalise the intercept?</em></p>
<p>The following code fits a group-regularized elastic net model (we set a seed, to make the results reproducible). Note that it may take a few minutes, depending on your computer.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
## estimate the group-regularized elastic net model
fitGrenCol &lt;-<span class="st"> </span><span class="kw">gren</span>(<span class="dt">x=</span>mirColScaled, <span class="dt">y=</span>respCol, <span class="dt">unpenalized=</span>unpenCol, 
                   <span class="dt">partitions=</span><span class="kw">list</span>(<span class="dt">expression=</span>mirExpr), <span class="dt">alpha=</span><span class="fl">0.5</span>)</code></pre></div>
</div>
<div id="interpreting-results" class="section level2">
<h2><span class="header-section-number">2.3</span> Interpreting results</h2>
<p>Next we look at the results. The estimated penalty multipliers may be retrieved from the object and plotted. We add a dotted line at one, to indicate the not group-regularized setting:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">barplot</span>(fitGrenCol<span class="op">$</span>lambdag<span class="op">$</span>expression, <span class="dt">main=</span><span class="st">&quot;Estimated penalty multipliers&quot;</span>,
        <span class="dt">ylab=</span><span class="kw">expression</span>(<span class="kw">paste</span>(lambda[g], <span class="st">&quot;'&quot;</span>)),
        <span class="dt">names.arg=</span><span class="kw">c</span>(<span class="st">&quot;Not expressed&quot;</span>, <span class="st">&quot;Medium differentially </span><span class="ch">\n</span><span class="st"> expressed&quot;</span>,
                    <span class="st">&quot;Highly differentially </span><span class="ch">\n</span><span class="st"> expressed&quot;</span>))
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">1</span>, <span class="dt">lty=</span><span class="dv">2</span>)</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>Why does the dotted line correspond to a regular (not group-regularized) model?</em></p>
<p><strong><em>QUESTION</em></strong> <em>Do the estimated penalty multipliers follow the expected pattern?</em></p>
<p>It might also be interesting to compare the model parameter estimates by the group-regularized regression with the regular elastic net estimates:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## extract the estimates from te fitted object
estRegular &lt;-<span class="st"> </span><span class="kw">coef</span>(fitGrenCol<span class="op">$</span>freq.model<span class="op">$</span>regular, <span class="dt">s=</span>fitGrenCol<span class="op">$</span>lambda)[, <span class="dv">1</span>]
estGroupreg &lt;-<span class="st"> </span><span class="kw">coef</span>(fitGrenCol<span class="op">$</span>freq.model<span class="op">$</span>groupreg, <span class="dt">s=</span>fitGrenCol<span class="op">$</span>lambda)[, <span class="dv">1</span>]

## plot the estimates in a scatterplot
<span class="kw">plot</span>(estRegular, estGroupreg, <span class="dt">col=</span>mirExpr, <span class="dt">xlab=</span><span class="kw">expression</span>(<span class="kw">hat</span>(beta)[regular]),
     <span class="dt">ylab=</span><span class="kw">expression</span>(<span class="kw">hat</span>(beta)[group]), 
     <span class="dt">main=</span><span class="st">&quot;Regular vs group-regularized estimates&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="dt">pch=</span><span class="dv">1</span>, <span class="dt">col=</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>),
       <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;Not expressed&quot;</span>, <span class="st">&quot;Medium differentially expressed&quot;</span>,
                <span class="st">&quot;Highly differentially expressed&quot;</span>))

<span class="kw">names</span>(<span class="kw">which</span>(estRegular<span class="op">!=</span><span class="dv">0</span>))
<span class="kw">names</span>(<span class="kw">which</span>(estGroupreg<span class="op">!=</span><span class="dv">0</span>))

<span class="kw">intersect</span>(<span class="kw">names</span>(estRegular)[<span class="kw">which</span>(estRegular<span class="op">!=</span><span class="dv">0</span>)[<span class="op">-</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>)]],
          <span class="kw">names</span>(estGroupreg)[<span class="kw">which</span>(estGroupreg<span class="op">!=</span><span class="dv">0</span>)[<span class="op">-</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>)]])</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>Which model selects more microRNAs, the group-regularized model or the regular model? What is the overlap between the two?</em></p>
<p>To reduce the costs of clinical predictions, it is often desirable to obtain a sparse final model. In addition, under equal model sizes, the comparison of two models is easier to interpret. Feature selection is possible by setting <code>psel</code> to the desired number of features and is done by adjusting the global lambda after the penalty multipliers have been estimated. Here, we select 50 features. Again, this will likely take two or three minutes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">2</span>)
fitGrenColSel &lt;-<span class="st"> </span><span class="kw">gren</span>(<span class="dt">x=</span>mirColScaled, <span class="dt">y=</span>respCol, <span class="dt">unpenalized=</span>unpenCol,
                      <span class="dt">partitions=</span><span class="kw">list</span>(<span class="dt">expression=</span>mirExpr), <span class="dt">alpha=</span><span class="fl">0.5</span>, <span class="dt">psel=</span><span class="dv">50</span>)

## extract the estimates from the model object
estRegularSel &lt;-<span class="st"> </span><span class="kw">coef</span>(fitGrenColSel<span class="op">$</span>freq.model<span class="op">$</span>regular, 
                      <span class="dt">s=</span>fitGrenColSel<span class="op">$</span>lambda)[, <span class="dv">1</span>]
estGroupregSel &lt;-<span class="st"> </span><span class="kw">coef</span>(fitGrenColSel<span class="op">$</span>freq.model<span class="op">$</span>groupreg, 
                       <span class="dt">s=</span>fitGrenColSel<span class="op">$</span>lambda)[, <span class="dv">1</span>]
## plot the estimates in a scatterplot
<span class="kw">plot</span>(estRegularSel, estGroupregSel, <span class="dt">col=</span>mirExpr, 
     <span class="dt">xlab=</span><span class="kw">expression</span>(<span class="kw">hat</span>(beta)[regular]), <span class="dt">ylab=</span><span class="kw">expression</span>(<span class="kw">hat</span>(beta)[group]), 
     <span class="dt">main=</span><span class="st">&quot;Regular vs group-regularized estimates, with model size 50&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="dt">pch=</span><span class="dv">1</span>, <span class="dt">col=</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>),
       <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;Not expressed&quot;</span>, <span class="st">&quot;Medium differentially expressed&quot;</span>,
                <span class="st">&quot;Highly differentially expressed&quot;</span>))

<span class="kw">names</span>(<span class="kw">which</span>(estRegularSel<span class="op">!=</span><span class="dv">0</span>))
<span class="kw">names</span>(<span class="kw">which</span>(estGroupregSel<span class="op">!=</span><span class="dv">0</span>))

<span class="kw">intersect</span>(<span class="kw">names</span>(estRegularSel)[<span class="kw">which</span>(estRegularSel<span class="op">!=</span><span class="dv">0</span>)[<span class="op">-</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>)]],
          <span class="kw">names</span>(estGroupregSel)[<span class="kw">which</span>(estGroupregSel<span class="op">!=</span><span class="dv">0</span>)[<span class="op">-</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>)]])</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>How did the results change compared to before? Is there more or less overlap between the selected microRNAs?</em></p>
</div>
<div id="predictive-performance" class="section level2">
<h2><span class="header-section-number">2.4</span> Predictive performance</h2>
<p>Since logistic regression is a classification method, we may compare classification performance of the regular and group-regularized elastic net. To avoid overestimation of performance we would like to use an indepenent data set to calculate the performance measures on. The simplest way of accomplishing this is by randomly splitting the data set in two parts, one for estimation and one for classification performance evaluation. But since we have a limited number of patients, splitting them in half would leave us with a very small dataset for estimation. With such a small data set we run the risk of bad estimators for the model. To avoid this we use cross-validation to estimate performance. Using cross-validation, we will still slightly overestimate performance, but since we are comparing two methods, we expect that this cancels out in the comparison.</p>
<p>The <code>gren</code> package contains a convenience function for performance cross-validation: <code>cv.gren</code>. This function is generally faster than naive cross-validation. An important extra argument is <code>fix.lambda</code>. If it is set to <code>TRUE</code>, the global lambda is fixed throughout the cross-validation iterations, which saves time, but slightly increases overestimation of performance. We compare with the regular elastic net by setting <code>compare=TRUE</code>. To make the comparison fair, we select the same number of features using <code>psel=50</code>.</p>
<p>Otherwise the following code estimates the performance by cross-validation (may take a few minutes to run):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">3</span>)
fitCvgrenCol &lt;-<span class="st"> </span><span class="kw">cv.gren</span>(<span class="dt">x=</span>mirColScaled, <span class="dt">y=</span>respCol, <span class="dt">alpha=</span><span class="fl">0.5</span>, 
                        <span class="dt">partitions=</span><span class="kw">list</span>(<span class="dt">expression=</span>mirExpr),
                        <span class="dt">unpenalized=</span>unpenCol, <span class="dt">psel=</span><span class="dv">50</span>, <span class="dt">nfolds.out=</span><span class="dv">10</span>)</code></pre></div>
<p>Due to time constraints, we used 10-fold cross-validation to estimate performance by setting <code>nfolds.out=10</code>. It is of course possible to use more folds by changing this value.</p>
<p>To assess predictive performance we compare the ROCs of the group-regularized elastic net with the regular elastic net.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## comparing ROCs for group-regularized and regular elastic net
<span class="kw">par</span>(<span class="dt">pty=</span><span class="st">&quot;s&quot;</span>)
<span class="kw">plot</span>(pROC<span class="op">::</span><span class="kw">roc</span>(respCol, <span class="kw">as.numeric</span>(fitCvgrenCol<span class="op">$</span>groupreg<span class="op">$</span>pred)), <span class="dt">print.auc=</span><span class="ot">TRUE</span>, 
     <span class="dt">print.auc.x=</span><span class="fl">0.3</span>, <span class="dt">print.auc.y=</span><span class="fl">0.2</span>)
<span class="kw">plot</span>(pROC<span class="op">::</span><span class="kw">roc</span>(respCol, <span class="kw">as.numeric</span>(fitCvgrenCol<span class="op">$</span>regular<span class="op">$</span>pred)), <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">col=</span><span class="dv">2</span>, 
     <span class="dt">print.auc=</span><span class="ot">TRUE</span>, <span class="dt">print.auc.x=</span><span class="fl">0.3</span>, <span class="dt">print.auc.y=</span><span class="fl">0.1</span>)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;Group-regularized&quot;</span>, <span class="st">&quot;Regular&quot;</span>), <span class="dt">col=</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>), <span class="dt">lty=</span><span class="dv">1</span>)
<span class="kw">par</span>(<span class="dt">pty=</span><span class="st">&quot;m&quot;</span>)</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>Do you think that the inclusion of the co-data helped with predictive performance here?</em></p>
<p>It might also be informative to plot the (cross-validated) predicted probabilites of the two models:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(<span class="kw">as.numeric</span>(fitCvgrenCol<span class="op">$</span>regular<span class="op">$</span>pred), 
     <span class="kw">as.numeric</span>(fitCvgrenCol<span class="op">$</span>groupreg<span class="op">$</span>pred), <span class="dt">col=</span>respCol, 
     <span class="dt">main=</span><span class="st">&quot;Predicted probabilities for Colon cancer data&quot;</span>,
     <span class="dt">xlab=</span><span class="st">&quot;Predictions regular model&quot;</span>, 
     <span class="dt">ylab=</span><span class="st">&quot;Predictions group-regularized model&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;progressive&quot;</span>, <span class="st">&quot;non-progressive/remission&quot;</span>), 
       <span class="dt">col=</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>), <span class="dt">pch=</span><span class="dv">1</span>)

<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">hist</span>(<span class="kw">as.numeric</span>(fitCvgrenCol<span class="op">$</span>groupreg<span class="op">$</span>pred), <span class="dt">main=</span><span class="st">&quot;Group-regularized&quot;</span>,
     <span class="dt">xlab=</span><span class="st">&quot;Predicted probabilities&quot;</span>)
<span class="kw">hist</span>(<span class="kw">as.numeric</span>(fitCvgrenCol<span class="op">$</span>regular<span class="op">$</span>pred), <span class="dt">main=</span><span class="st">&quot;Not group-regularized&quot;</span>,
     <span class="dt">xlab=</span><span class="st">&quot;Predicted probabilities&quot;</span>)
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>In what way does the inclusion of the co-data change the predicted probabilities?</em></p>
</div>
</div>
<div id="example-2-micrornas-in-cervical-cancer" class="section level1">
<h1><span class="header-section-number">3</span> Example 2: microRNAs in cervical cancer</h1>
<div id="data-1" class="section level2">
<h2><span class="header-section-number">3.1</span> Data</h2>
<p>In this study, microRNAs from women were collected through self-samples. These women were either cervical cancer disease free or with high grade cervical lesions (CIN2/3). These lesions have a high risk for progression to cancer and hence need accurate detection <span class="citation">(Novianti et al. <a href="#ref-novianti_better_2017">2017</a>)</span>. The data consists of (transformed) sequenced microRNAs. We first load the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## load the data and inspect the different data object
<span class="kw">data</span>(dataCervical)</code></pre></div>
<p>The mircoRNAs are given as a <code>matrix</code> object named <code>mirCerv</code>. The dependent variable is in the form of a <code>factor</code> called <code>respCerv</code>. In addition, there is the <code>mirCons</code> object, which contains the conservation status of the microRNAs <span class="citation">(Agarwal et al. <a href="#ref-agarwal_predicting_nodate">2018</a>)</span>. The conservation status comes in three levels: not conserved, conserved across most mammals, and broadly conserved across most vertebrates. We may include conservation status as co-data, since we expect that conserved microRNAs are more important for body functioning and consequently might be important in CIN2/3 development.</p>
</div>
<div id="two-partitions" class="section level2">
<h2><span class="header-section-number">3.2</span> Two partitions</h2>
<p>We may create additional co-data by examining the microRNAs themselves. Here we create a co-data set based on the sample standard deviations of the microRNAs:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## create 5 equally sized groups based on the standard deviations
std &lt;-<span class="st"> </span><span class="kw">apply</span>(mirCerv, <span class="dv">2</span>, sd)
mirStd &lt;-<span class="st"> </span><span class="kw">cut</span>(std, <span class="dt">breaks=</span><span class="kw">quantile</span>(std, <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.2</span>)), <span class="dt">include.lowest=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>What do you think the relation between the standard deviation and penalty parameter of the microRNAs should be?</em></p>
<p>We may include this co-data in the estimation by extending the <code>partitions</code> argument:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## create a partitions objected that contains the different co-data sets
partitions &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">conservation=</span>mirCons, <span class="dt">std=</span>mirStd)</code></pre></div>
<p>We may also specify an extra argument <code>monotone</code>, a <code>list</code> containing two <code>logical</code> vectors. The first indicates the whether the corresponding partitions penalty multipliers are monotone, while the second indicates whether the corresponding parititions penalty multipliers are monotonically decreasing with group number or not</p>
<p><strong><em>QUESTION</em></strong> <em>Would you impose a monotonicity constraint on the standard deviations partition? And if so, in what direction would the monotonicity be?</em></p>
<p>To include a monotonicity constraint create the following object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## create a partitions objected that contains the different co-data sets
monotone &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">c</span>(<span class="ot">FALSE</span>, <span class="ot">TRUE</span>), <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">TRUE</span>))</code></pre></div>
<p>To estimate the model with the monotonicity constraint, <code>alpha=0.5</code>, and with the selection of 50 features, run the following chunk of code (this may take a few minutes):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## prepare the data and estimate a model
mirCervScaled &lt;-<span class="st"> </span><span class="kw">scale</span>(mirCerv)

<span class="kw">set.seed</span>(<span class="dv">4</span>)
fitGrenCervSel &lt;-<span class="st"> </span><span class="kw">gren</span>(<span class="dt">x=</span>mirCervScaled, <span class="dt">y=</span>respCerv, <span class="dt">partitions=</span>partitions, 
                       <span class="dt">alpha=</span><span class="fl">0.5</span>, <span class="dt">monotone=</span>monotone, <span class="dt">psel=</span><span class="dv">50</span>)</code></pre></div>
<p>Again, we may inspect the penalty multipliers. This time we have two sets, so we create two barplots:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">barplot</span>(fitGrenCervSel<span class="op">$</span>lambdag<span class="op">$</span>conservation, 
        <span class="dt">main=</span><span class="st">&quot;Penalty multipliers conservation&quot;</span>, 
        <span class="dt">ylab=</span><span class="kw">expression</span>(<span class="kw">paste</span>(lambda[g], <span class="st">&quot;'&quot;</span>)),
        <span class="dt">names.arg=</span><span class="kw">c</span>(<span class="st">&quot;Not conserved&quot;</span>, <span class="st">&quot;Conserved in mammals&quot;</span>,
                    <span class="st">&quot;Broadly conserved </span><span class="ch">\n</span><span class="st"> across vertebrates&quot;</span>))
<span class="kw">barplot</span>(fitGrenCervSel<span class="op">$</span>lambdag<span class="op">$</span>std, 
        <span class="dt">main=</span><span class="st">&quot;Penalty multipliers standard deviation&quot;</span>, 
        <span class="dt">ylab=</span><span class="kw">expression</span>(<span class="kw">paste</span>(lambda[g], <span class="st">&quot;'&quot;</span>)),
        <span class="dt">names.arg=</span><span class="kw">levels</span>(mirStd))
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>Are the penalty multipliers of the conservation status according to expectation? And what about the standard deviation penalty multipliers?</em></p>
<p>Since we now have two partitions, it is not straightforward to see what the total penalty multiplication for each feature is. We calculate and inspect the total penalty multiplication per feature by:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## create a vector with the penalty multipliers per feature in it
multvec &lt;-<span class="st"> </span>fitGrenCervSel<span class="op">$</span>lambdag<span class="op">$</span>conservation[
  <span class="kw">match</span>(mirCons, <span class="kw">names</span>(fitGrenCervSel<span class="op">$</span>lambdag<span class="op">$</span>conservation))]<span class="op">*</span>
<span class="st">  </span>fitGrenCervSel<span class="op">$</span>lambdag<span class="op">$</span>std[<span class="kw">match</span>(mirStd, <span class="kw">names</span>(fitGrenCervSel<span class="op">$</span>lambdag<span class="op">$</span>std))]

## and create a histogram of the multipliers
<span class="kw">hist</span>(multvec, <span class="dt">main=</span><span class="st">&quot;Total penalty multiplier per feature&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Penalty multiplier&quot;</span>)</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>Can you discern the different groups from the two partitions from the histogram? Why/why not?</em></p>
<p>Again, we may use cross validation to estimate predictive performance of the model. The following code estimates the performance by cross-validation (may take a few minutes to estimate):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">5</span>)
fitCvgrenCerv &lt;-<span class="st"> </span><span class="kw">cv.gren</span>(<span class="dt">x=</span>mirCervScaled, <span class="dt">y=</span>respCerv, <span class="dt">alpha=</span><span class="fl">0.5</span>, 
                         <span class="dt">partitions=</span>partitions, <span class="dt">monotone=</span>monotone, <span class="dt">psel=</span><span class="dv">50</span>, 
                         <span class="dt">nfolds.out=</span><span class="dv">10</span>)</code></pre></div>
<p>Again, we look at ROC plots to compare the performance of the group-regularized elastic net with the non group-regularized elastic net:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## comparing ROCs for group-regularized and regular elastic net
<span class="kw">par</span>(<span class="dt">pty=</span><span class="st">&quot;s&quot;</span>)
<span class="kw">plot</span>(pROC<span class="op">::</span><span class="kw">roc</span>(respCerv, <span class="kw">as.numeric</span>(fitCvgrenCerv<span class="op">$</span>groupreg<span class="op">$</span>pred)), 
     <span class="dt">print.auc=</span><span class="ot">TRUE</span>, <span class="dt">print.auc.x=</span><span class="fl">0.3</span>, <span class="dt">print.auc.y=</span><span class="fl">0.2</span>)
<span class="kw">plot</span>(pROC<span class="op">::</span><span class="kw">roc</span>(respCerv, <span class="kw">as.numeric</span>(fitCvgrenCerv<span class="op">$</span>regular<span class="op">$</span>pred)), <span class="dt">add=</span><span class="ot">TRUE</span>, 
     <span class="dt">col=</span><span class="dv">2</span>, <span class="dt">print.auc=</span><span class="ot">TRUE</span>, <span class="dt">print.auc.x=</span><span class="fl">0.3</span>, <span class="dt">print.auc.y=</span><span class="fl">0.1</span>)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;Group-regularized&quot;</span>, <span class="st">&quot;Regular&quot;</span>), <span class="dt">col=</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>), <span class="dt">lty=</span><span class="dv">1</span>)
<span class="kw">par</span>(<span class="dt">pty=</span><span class="st">&quot;m&quot;</span>)</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>Do you think that the inclusion of the co-data helped with predictive performance here?</em></p>
<p>It might also be informative to plot the (cross-validated) predicted probabilites of the two models:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(<span class="kw">as.numeric</span>(fitCvgrenCerv<span class="op">$</span>regular<span class="op">$</span>pred), 
     <span class="kw">as.numeric</span>(fitCvgrenCerv<span class="op">$</span>groupreg<span class="op">$</span>pred), <span class="dt">col=</span>respCerv, 
     <span class="dt">main=</span><span class="st">&quot;Predicted probabilities for Cervical cancer data&quot;</span>,
     <span class="dt">xlab=</span><span class="st">&quot;Predictions regular model&quot;</span>, 
     <span class="dt">ylab=</span><span class="st">&quot;Predictions group-regularized model&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;CIN3&quot;</span>, <span class="st">&quot;normal&quot;</span>), 
       <span class="dt">col=</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>), <span class="dt">pch=</span><span class="dv">1</span>)

<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">hist</span>(<span class="kw">as.numeric</span>(fitCvgrenCerv<span class="op">$</span>groupreg<span class="op">$</span>pred), <span class="dt">main=</span><span class="st">&quot;Group-regularized&quot;</span>,
     <span class="dt">xlab=</span><span class="st">&quot;Predicted probabilities&quot;</span>)
<span class="kw">hist</span>(<span class="kw">as.numeric</span>(fitCvgrenCerv<span class="op">$</span>regular<span class="op">$</span>pred), <span class="dt">main=</span><span class="st">&quot;Not group-regularized&quot;</span>,
     <span class="dt">xlab=</span><span class="st">&quot;Predicted probabilities&quot;</span>)
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</code></pre></div>
<p><strong><em>QUESTION</em></strong> <em>In what way does the inclusion of the co-data change the predicted probabilities?</em></p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references">
<div id="ref-agarwal_predicting_nodate">
<p>Agarwal, Vikram, George W Bell, Jin-Wu Nam, and David P Bartel. 2018. âPredicting Effective microRNA Target Sites in Mammalian mRNAs.â <em>eLife</em> 4. Accessed July 19. doi:<a href="https://doi.org/10.7554/eLife.05005">10.7554/eLife.05005</a>.</p>
</div>
<div id="ref-munch_adaptive_2018">
<p>MÃ¼nch, Magnus M., Carel F. W. Peeters, Aad W. van der Vaart, and Mark A. van de Wiel. 2018. âAdaptive Group-Regularized Logistic Elastic Net Regression.â <em>arXiv:1805.00389 [Stat]</em>, May. <a href="http://arxiv.org/abs/1805.00389" class="uri">http://arxiv.org/abs/1805.00389</a>.</p>
</div>
<div id="ref-neerincx_mir_2015">
<p>Neerincx, M., D. L. S. Sie, M. A. van de Wiel, N. C. T. van Grieken, J. D. Burggraaf, H. Dekker, P. P. Eijk, et al. 2015. âMiR Expression Profiles of Paired Primary Colorectal Cancer and Metastases by Next-Generation Sequencing.â <em>Oncogenesis</em> 4 (10): e170. doi:<a href="https://doi.org/10.1038/oncsis.2015.29">10.1038/oncsis.2015.29</a>.</p>
</div>
<div id="ref-novianti_better_2017">
<p>Novianti, P. W., B. C. Snoek, S. M. Wilting, and M. A. van de Wiel. 2017. âBetter Diagnostic Signatures from RNAseq Data Through Use of Auxiliary Co-Data.â <em>Bioinformatics</em> 30 (10): 1572â4. doi:<a href="https://doi.org/10.1093/bioinformatics/btw837">10.1093/bioinformatics/btw837</a>.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
